packages/suite/src/middlewares/wallet/coinjoinMiddleware.ts

```diff
@@ -164,15 +88,77 @@ export const coinjoinMiddleware =
+                api.dispatch(coinjoinAccountActions.pauseInterruptAllCoinjoinSessions());
+            } else if (action.payload === true) {
+                api.dispatch(coinjoinAccountActions.restoreInterruptedCoinjoinSessions());
+            }
+        }
+
+        // Pause/restore coinjoin session based on Tor status.
+        // Continuing coinjoin would be a privacy risk.
         if (action.type === SUITE.TOR_STATUS) {
             if (['Disabling', 'Disabled', 'Error'].includes(action.payload)) {
                 api.dispatch(coinjoinAccountActions.pauseInterruptAllCoinjoinSessions());
+            } else if (action.payload === 'Enabled') {
+                api.dispatch(coinjoinAccountActions.restoreInterruptedCoinjoinSessions());
             }
-            // We restore sessions that were interrupted when successfully Enabled if
-            // there is not any other condition blocking coinjoin resume.
+        }
+
+        // Pause/restore coinjoin session when an account goes out of sync or in sync.
+        // As this is not crucial, it does not pause during the critical phase not to ruin a round.
+        if (accountsActions.endCoinjoinAccountSync.match(action)) {
             const state = api.getState();
-            if (action.payload === 'Enabled' && !isCoinjoinSessionBlockedGlobally(state)) {
-                restoreInterruptedCoinjoinSessions(state);
+            const { accountKey, status } = action.payload;
+            const session = selectCoinjoinAccountByKey(state, accountKey)?.session;
+            if (status === 'out-of-sync' && session && !session?.paused && !session?.starting) {
+                const isAccountInCriticalPhase =
+                    selectIsAccountWithSessionInCriticalPhaseByAccountKey(state, accountKey);
+                if (!isAccountInCriticalPhase) {
+                    api.dispatch(coinjoinAccountActions.pauseCoinjoinSession(accountKey, true));
+                }
+            } else if (status === 'ready' && session?.interrupted) {
+                const account = selectAccountByKey(state, accountKey);
+                if (account) {
+                    const blockers = selectCoinjoinSessionBlockersByAccount(state, account);
+                    if (!Object.values(blockers).some(value => value))
+                        api.dispatch(coinjoinAccountActions.restoreCoinjoinSession(account.key));
+                }
+            }
+        }
+
+        // Pause/restore coinjoin session depending on current route.
+        // Device may be locked by another connect call, so check on LOCK_DEVICE action as well.
+        if (action.type === ROUTER.LOCATION_CHANGE || action.type === SUITE.LOCK_DEVICE) {
+            const state = api.getState();
+            const { locks } = state.suite;
+            if (!locks.includes(SUITE.LOCK_TYPE.DEVICE) && !locks.includes(SUITE.LOCK_TYPE.UI)) {
+                if (state.router.settingsBackRoute.name === 'wallet-send') {
```

Comment:
you are checking for the previous route here, right?

Reply:
right

Reply:
I know it's unrelated but that router param would definitely benefit from renaming 😓

Reply:
But `settingsBackRoute` is meant to be used as back route from settings when you click X button in settings so all settings routes are excluded and routes of foreground apps (modals) too. Not sure if it's ok to reuse it here, but maybe it is.

Reply:
Pardon, of course, rushed with my answer. It just looks weird here, I guess that's what I wanted to say :D 

**nit:** could be helpful to recontextualize it with:

```
const isNavigatingAwayFromSend = state.router.settingsBackRoute.name === 'wallet-send'
```

Reply:
Recontextualized in 0dee27a7e4512783a15c46301b1be41104f2e163.